### PR TITLE
Use face_no in FE

### PR DESCRIPTION
**Pull request by peterrum**
~~Only the last commit is relevant.~~

### CODE REVIEW

File: source/fe/fe_trace.cc [hunk 1]
```diff
@@ -53,7 +53,10 @@ FE_TraceQ<dim, spacedim>::FE_TraceQ(const unsigned int degree)
     FETools::hierarchic_to_lexicographic_numbering<dim - 1>(degree));
 
   // Initialize face support points
-  this->unit_face_support_points[0] = fe_q.get_unit_face_support_points();
+  AssertDimension(this->n_unique_faces(), fe_q.n_unique_faces());
+  for (unsigned int face_no; face_no < this->n_unique_faces(); ++face_no)
```
Comment (masterleinad) [inline source/fe/fe_trace.cc:57]:
```suggestion
  for (unsigned int face_no = 0; face_no < this->n_unique_faces(); ++face_no)
```

File: source/fe/fe_nedelec.cc [hunk 2]
```diff
@@ -2453,20 +2470,20 @@ void
 FE_Nedelec<dim>::get_face_interpolation_matrix(
   const FiniteElement<dim> &source,
   FullMatrix<double> &      interpolation_matrix,
-  const unsigned int) const
+  const unsigned int        face_no) const
```
Comment (masterleinad) [inline source/fe/fe_nedelec.cc:2473]: This is only used in `Debug` mode.

Comment (peterrum) [reply source/fe/fe_nedelec.cc:2473]: Should be fine now!

File: include/deal.II/fe/fe_base.h [hunk 3]
```diff
@@ -592,7 +592,8 @@ class FiniteElementData
    */
   template <int structdim>
   unsigned int
-  n_dofs_per_object(const unsigned int i = 0) const;
+  n_dofs_per_object(
+    const unsigned int i = 0 /*numbers::invalid_unsigned_int*/) const;
```
Comment (masterleinad) [inline include/deal.II/fe/fe_base.h:596]: What is this about?

File: include/deal.II/fe/fe_base.h [hunk 4]
```diff
@@ -840,6 +841,10 @@ FiniteElementData<dim>::n_dofs_per_object(const unsigned int i) const
       case 1:
         return n_dofs_per_line();
       case 2:
+        // Assert(!(structdim == 2 && dim == 3) || i !=
+        // numbers::invalid_unsigned_int,
+        //         ExcInternalError());
```
Comment (masterleinad) [inline include/deal.II/fe/fe_base.h:846]: ???

File: source/fe/fe_system.cc [hunk 5]
```diff
@@ -1702,43 +1719,52 @@ FESystem<dim, spacedim>::initialize(
...
+          // degrees of freedom on the face at all, then we don't care whether
+          // it has support points or not. this is, for example, the case for
+          // the stable Stokes element Q(p)^dim \times DGP(p-1).
+          for (unsigned int base_el = 0; base_el < this->n_base_elements();
+               ++base_el)
+            if (!base_element(base_el).has_support_points() &&
+                (base_element(base_el).n_dofs_per_face(face_no) > 0))
+              {
+                this->unit_face_support_points[face_no].resize(0);
+                goto face_loop;
```
Comment (masterleinad) [inline source/fe/fe_system.cc:1741]: really?! Let's not do this. Maybe define another boolean instead and `continue` after the inner loop in the outer loop.

Comment (peterrum) [reply source/fe/fe_system.cc]: Done!

File: include/deal.II/fe/fe_base.h (hunk 3 above)
Comment (peterrum) [reply include/deal.II/fe/fe_base.h]: This and the next was for "debugging". I have remove them.

File: include/deal.II/fe/fe.h [hunk 6]
```diff
@@ -1469,7 +1469,7 @@ class FiniteElement : public Subscriptor, public FiniteElementData<dim>
    */
   std::pair<unsigned int, unsigned int>
   face_system_to_component_index(const unsigned int index,
-                                 const unsigned int face_no = 0) const;
+                                 const unsigned int face = 0) const;
```
Comment (kronbichler) [inline include/deal.II/fe/fe.h:1472]: Is this change intentional? It seems to contract the other `face_no` argument names?

File: include/deal.II/fe/fe.h [hunk 7]
```diff
@@ -1481,7 +1481,7 @@ class FiniteElement : public Subscriptor, public FiniteElementData<dim>
    */
   unsigned int
   adjust_quad_dof_index_for_face_orientation(const unsigned int index,
-                                             const unsigned int face_no,
+                                             const unsigned int face,
```
Comment (kronbichler) [inline include/deal.II/fe/fe.h:1484]: Same here.

File: include/deal.II/fe/fe.h [hunk 8]
```diff
@@ -1766,7 +1766,7 @@ class FiniteElement : public Subscriptor, public FiniteElementData<dim>
    */
   std::pair<std::pair<unsigned int, unsigned int>, unsigned int>
   face_system_to_base_index(const unsigned int index,
-                            const unsigned int face_no = 0) const;
+                            const unsigned int face_no) const;
```
Comment (kronbichler) [inline include/deal.II/fe/fe.h:1769]: Do we need to document this change somewhere as this can have user-visible effects?

File: include/deal.II/fe/fe.h (hunk 6 above)
Comment (peterrum) [reply include/deal.II/fe/fe.h]: I have reverted this change.

File: include/deal.II/fe/fe.h (hunk 8 above)
Comment (peterrum) [reply include/deal.II/fe/fe.h]: This was not intended. I have removed the default value to be able to identify all instances where these functions are used. Reverted!

Comment (kronbichler) [reply include/deal.II/fe/fe.h]: > This was not intended.

That's why we have code review :wink:
